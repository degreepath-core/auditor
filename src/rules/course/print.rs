use super::*;
use crate::traits::print;
use crate::util;

impl print::Print for Rule {
	fn print(&self) -> print::Result {
		use std::fmt::Write;
		let mut output = String::new();

		write!(&mut output, "{}", self.course)?;

		if let Some(section) = &self.section {
			write!(&mut output, "{}", section)?;
		}

		let mut annotations = match (self.year, &self.semester) {
			(Some(year), None) => Some(util::expand_year(year, "dual")),
			(None, Some(semester)) => Some(semester.to_string()),
			(Some(year), Some(semester)) => Some(format!("{} {}", semester, util::expand_year(year, "short"))),
			(None, None) => None,
		};

		if let Some(lab) = self.lab {
			if lab {
				match &annotations {
					Some(ann) => {
						annotations = Some(format!("Lab; {}", ann));
					}
					None => {
						annotations = Some("Lab".to_string());
					}
				}
			}
		}

		match (&self.grade, &annotations) {
			(Some(grade), Some(ann)) => annotations = Some(format!("{}; {} or higher", ann, String::from(grade))),
			(Some(grade), None) => annotations = Some(format!("{} or higher", String::from(grade))),
			_ => (),
		}

		if let Some(annotations) = annotations {
			write!(&mut output, " ({})", annotations)?;
		}

		Ok(output)
	}
}
