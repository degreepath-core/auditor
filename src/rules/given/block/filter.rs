#[derive(Debug, PartialEq, Serialize, Deserialize, Clone)]
pub struct Clause {
    pub key: String,
    pub value: String,
    pub operation: Op,
}

#[derive(Debug, PartialEq, Serialize, Deserialize, Clone)]
pub enum Op {
    Eq,
    NotEq,
}
