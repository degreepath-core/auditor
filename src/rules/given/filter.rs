#[derive(Debug, PartialEq, Serialize, Deserialize, Clone)]
pub struct Clause {
    pub key: String,
    pub value: String,
    pub operation: Operator,
}

#[derive(Debug, PartialEq, Serialize, Deserialize, Clone)]
pub enum Operator {
    EqualTo,
    NotEqualTo,
}
