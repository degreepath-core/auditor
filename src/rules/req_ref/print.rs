use super::*;
use crate::traits::print;

impl print::Print for Rule {
	fn print(&self) -> print::Result {
		use std::fmt::Write;
		let mut output = String::new();

		write!(&mut output, "“{}”", self.name)?;

		if self.optional {
			write!(&mut output, " (optional)")?;
		}

		Ok(output)
	}
}
