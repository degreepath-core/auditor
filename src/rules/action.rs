use crate::rules::given::action;
use crate::util;

#[derive(Debug, PartialEq, Serialize, Deserialize, Clone)]
pub struct Rule {
    #[serde(rename = "do", deserialize_with = "util::string_or_struct_parseerror")]
    pub action: action::Action,
}
