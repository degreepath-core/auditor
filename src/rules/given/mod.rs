use crate::filter;
use crate::limit;
use crate::rules::course;
use crate::traits::Util;
use crate::util;
use crate::value::WrappedValue;
use decorum::R32;
use serde::{Deserialize, Serialize};

// mod audit;
mod print;
#[cfg(test)]
mod tests;

#[derive(Debug, PartialEq, Serialize, Deserialize, Clone)]
#[serde(tag = "given", rename_all = "kebab-case")]
pub enum Rule {
	#[serde(rename = "courses")]
	AllCourses {
		what: GivenCoursesWhatOptions,
		#[serde(rename = "where", default)]
		filter: Option<filter::CourseClause>,
		#[serde(default)]
		limit: Option<Vec<limit::Limiter>>,
	},
	TheseCourses {
		courses: Vec<CourseRule>,
		repeats: RepeatMode,
		what: GivenCoursesWhatOptions,
		#[serde(rename = "where", default)]
		filter: Option<filter::CourseClause>,
		#[serde(default)]
		limit: Option<Vec<limit::Limiter>>,
	},
	TheseRequirements {
		requirements: Vec<String>,
		what: GivenCoursesWhatOptions,
		#[serde(rename = "where", default)]
		filter: Option<filter::CourseClause>,
		#[serde(default)]
		limit: Option<Vec<limit::Limiter>>,
	},
	Areas {
		what: GivenAreasWhatOptions,
		#[serde(rename = "where", default)]
		filter: Option<filter::AreaClause>,
	},
	#[serde(rename = "save")]
	NamedVariable {
		save: String,
		what: GivenCoursesWhatOptions,
		#[serde(rename = "where", default)]
		filter: Option<filter::CourseClause>,
		#[serde(default)]
		limit: Option<Vec<limit::Limiter>>,
	},
	Performances {
		what: GivenPerformancesWhatOptions,
		#[serde(rename = "where", default)]
		filter: Option<filter::PerformanceClause>,
	},
	Attendances {
		what: GivenAttendancesWhatOptions,
		#[serde(rename = "where", default)]
		filter: Option<filter::AttendanceClause>,
	},
}

impl Util for Rule {
	fn has_save_rule(&self) -> bool {
		match &self {
			Rule::NamedVariable { .. } => true,
			_ => false,
		}
	}
}

#[derive(Debug, PartialEq, Serialize, Deserialize, Clone)]
#[serde(tag = "given", rename_all = "kebab-case")]
pub enum GivenForSaveBlock {
	#[serde(rename = "courses")]
	AllCourses { what: GivenCoursesWhatOptions },
	TheseCourses {
		courses: Vec<CourseRule>,
		repeats: RepeatMode,
		what: GivenCoursesWhatOptions,
	},
	TheseRequirements {
		requirements: Vec<String>,
		what: GivenCoursesWhatOptions,
	},
	#[serde(rename = "save")]
	NamedVariable {
		save: String,
		what: GivenCoursesWhatOptions,
	},
}

use util::string_or_struct_parseerror as parse;

#[derive(Debug, PartialEq, Serialize, Deserialize, Clone)]
#[serde(rename_all = "kebab-case")]
pub enum GivenCoursesWhatOptions {
	Courses {
		#[serde(flatten, deserialize_with = "option_t")]
		action: Option<CountOnlyAction>,
	},
	DistinctCourses {
		#[serde(flatten, deserialize_with = "option_t")]
		action: Option<CountOnlyAction>,
	},
	Credits {
		#[serde(flatten, deserialize_with = "option_t")]
		action: Option<AnyAction>,
	},
	Subjects {
		#[serde(flatten, deserialize_with = "option_t")]
		action: Option<CountOnlyAction>,
	},
	Terms {
		#[serde(flatten, deserialize_with = "option_t")]
		action: Option<CountOrOrdAction>,
	},
	Grades {
		#[serde(flatten, deserialize_with = "option_t")]
		action: Option<AnyAction>,
	},
}

pub fn option_t<'de, D, T>(deserializer: D) -> Result<Option<T>, D::Error>
where
	D: serde::de::Deserializer<'de>,
	T: serde::de::Deserialize<'de>,
{
	#[derive(Deserialize)]
	struct Wrapper<T>(T);

	let v = Option::deserialize(deserializer)?;
	Ok(v.map(|Wrapper(a)| a))
}

#[derive(Debug, Eq, PartialEq, Serialize, Deserialize, Clone)]
#[serde(rename_all = "kebab-case")]
pub enum CountOnlyAction {
	Count(#[serde(deserialize_with = "parse")] WrappedValue<u64>),
}

impl crate::util::Pluralizable for CountOnlyAction {
	fn should_pluralize(&self) -> bool {
		use CountOnlyAction::*;
		match &self {
			Count(v) => v.should_pluralize(),
		}
	}
}

#[derive(Debug, Eq, PartialEq, Serialize, Deserialize, Clone)]
#[serde(rename_all = "kebab-case")]
pub enum CountOrOrdAction {
	Count(#[serde(deserialize_with = "parse")] WrappedValue<u64>),
	Minimum,
	Maximum,
}

impl crate::util::Pluralizable for CountOrOrdAction {
	fn should_pluralize(&self) -> bool {
		use CountOrOrdAction::*;
		match &self {
			Count(v) => v.should_pluralize(),
			Maximum => false,
			Minimum => false,
		}
	}
}

#[derive(Debug, Eq, PartialEq, Serialize, Deserialize, Clone)]
#[serde(rename_all = "kebab-case")]
pub enum AnyAction {
	Count(#[serde(deserialize_with = "parse")] WrappedValue<u64>),
	Sum(#[serde(deserialize_with = "parse")] WrappedValue<R32>),
	Average(#[serde(deserialize_with = "parse")] WrappedValue<R32>),
	Minimum,
	Maximum,
}

impl crate::util::Pluralizable for AnyAction {
	fn should_pluralize(&self) -> bool {
		use AnyAction::*;
		match &self {
			Count(v) => v.should_pluralize(),
			Sum(v) => v.should_pluralize(),
			Average(v) => v.should_pluralize(),
			Maximum => false,
			Minimum => false,
		}
	}
}

mod printable {
	use super::{AnyAction, CountOnlyAction, CountOrOrdAction};
	use crate::traits::print;
	use std::fmt::Write;

	impl print::Print for AnyAction {
		fn print(&self) -> print::Result {
			use crate::value::{TaggedValue, WrappedValue};
			use AnyAction::*;
			let mut output = String::new();

			match &self {
				Average(WrappedValue::Single(TaggedValue::GreaterThan(v))) => {
					write!(&mut output, "above {}", v.print()?)?
				}
				Average(WrappedValue::Single(TaggedValue::GreaterThanEqualTo(v))) => {
					write!(&mut output, "at or above {}", v.print()?)?
				}
				Count(v) => write!(&mut output, "{}", v.print()?)?,
				Sum(v) => write!(&mut output, "{}", v.print()?)?,
				Average(v) => write!(&mut output, "{}", v.print()?)?,
				Minimum => write!(&mut output, "the smallest item [todo]")?,
				Maximum => write!(&mut output, "the smallest item [todo]")?,
			}

			Ok(output)
		}
	}

	impl print::Print for CountOrOrdAction {
		fn print(&self) -> print::Result {
			use CountOrOrdAction::*;
			let mut output = String::new();

			match &self {
				Count(v) => write!(&mut output, "{}", v.print()?)?,
				Minimum => write!(&mut output, "the smallest item [todo]")?,
				Maximum => write!(&mut output, "the smallest item [todo]")?,
			}

			Ok(output)
		}
	}

	impl print::Print for CountOnlyAction {
		fn print(&self) -> print::Result {
			use CountOnlyAction::*;
			let mut output = String::new();

			match &self {
				Count(v) => write!(&mut output, "{}", v.print()?)?,
			}

			Ok(output)
		}
	}
}

#[derive(Debug, PartialEq, Serialize, Deserialize, Clone)]
#[serde(untagged)]
pub enum CourseRule {
	Value(#[serde(deserialize_with = "util::string_or_struct")] course::Rule),
}

impl crate::traits::print::Print for CourseRule {
	fn print(&self) -> crate::traits::print::Result {
		match &self {
			CourseRule::Value(v) => v.print(),
		}
	}
}

#[derive(Debug, Eq, PartialEq, Serialize, Deserialize, Clone)]
#[serde(rename_all = "kebab-case")]
pub enum RepeatMode {
	First,
	Last,
	All,
}

#[derive(Debug, Eq, PartialEq, Serialize, Deserialize, Clone)]
#[serde(rename_all = "kebab-case")]
pub enum GivenAreasWhatOptions {
	Areas {
		#[serde(flatten, deserialize_with = "option_t")]
		action: Option<CountOnlyAction>,
	},
}

#[derive(Debug, Eq, PartialEq, Serialize, Deserialize, Clone)]
#[serde(rename_all = "kebab-case")]
pub enum GivenPerformancesWhatOptions {
	Performances {
		#[serde(flatten, deserialize_with = "option_t")]
		action: Option<CountOrOrdAction>,
	},
}

#[derive(Debug, Eq, PartialEq, Serialize, Deserialize, Clone)]
#[serde(rename_all = "kebab-case")]
pub enum GivenAttendancesWhatOptions {
	Attendances {
		#[serde(flatten, deserialize_with = "option_t")]
		action: Option<CountOrOrdAction>,
	},
}
