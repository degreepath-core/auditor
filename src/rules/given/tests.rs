use super::Rule;
use super::{CourseRule, Given, GivenAreasWhatOptions, GivenCoursesWhatOptions, RepeatMode};
use crate::filter;
use crate::rules::{course, req_ref};
use crate::traits::print::Print;
use crate::value;
use insta::{assert_snapshot_matches, assert_debug_snapshot_matches};
use maplit::btreemap;

#[test]
fn serialize_all_courses() {
	let data = Rule {
		given: Given::AllCourses {
			what: GivenCoursesWhatOptions::Courses,
		},
		limit: Some(vec![]),
		filter: Some(filter::Clause::new()),
		action: "count > 2".parse().unwrap(),
	};

	assert_debug_snapshot_matches!("ser_all_courses", data);
}

#[test]
fn deserialize_all_courses() {
	let expected = Rule {
		given: Given::AllCourses {
			what: GivenCoursesWhatOptions::Courses,
		},
		limit: Some(vec![]),
		filter: Some(filter::Clause::new()),
		action: "count > 2".parse().unwrap(),
	};

	let data = r#"---
given: courses
what: courses
limit: []
where: {}
do: count > 2"#;

	let actual: Rule = serde_yaml::from_str(&data).unwrap();

	assert_eq!(actual, expected);

	let data = r#"---
given: courses
what: courses
limit: []
where: {}
do:
  lhs: Count
  op: GreaterThan
  rhs:
    Integer: 2"#;

	let actual: Rule = serde_yaml::from_str(&data).unwrap();

	assert_eq!(actual, expected);
}

#[test]
fn serialize_these_courses() {
	let data = Rule {
		given: Given::TheseCourses {
			courses: vec![
				CourseRule::Value(course::Rule {
					course: "ASIAN 110".to_string(),
					..Default::default()
				}),
				CourseRule::Value(course::Rule {
					course: "ASIAN 110".to_string(),
					..Default::default()
				}),
			],
			repeats: RepeatMode::First,
			what: GivenCoursesWhatOptions::Courses,
		},
		limit: Some(vec![]),
		filter: Some(filter::Clause::new()),
		action: "count > 2".parse().unwrap(),
	};

	assert_debug_snapshot_matches!("ser_these_courses_1", data);
}

#[test]
fn deserialize_these_courses() {
	let expected = Rule {
		given: Given::TheseCourses {
			courses: vec![
				CourseRule::Value(course::Rule {
					course: "ASIAN 110".to_string(),
					..Default::default()
				}),
				CourseRule::Value(course::Rule {
					course: "ASIAN 110".to_string(),
					..Default::default()
				}),
			],
			repeats: RepeatMode::First,
			what: GivenCoursesWhatOptions::Courses,
		},
		limit: Some(vec![]),
		filter: Some(filter::Clause::new()),
		action: "count > 2".parse().unwrap(),
	};

	let data = r#"---
given: these-courses
courses:
  - ASIAN 110
  - course: ASIAN 110
repeats: first
what: courses
limit: []
where: {}
do: count > 2"#;

	let actual: Rule = serde_yaml::from_str(&data).unwrap();

	assert_eq!(actual, expected);

	let data = r#"---
given: these-courses
courses:
  - ASIAN 110
  - course: ASIAN 110
repeats: first
what: courses
limit: []
where: {}
do:
  lhs: Count
  op: GreaterThan
  rhs:
    Integer: 2"#;

	let actual: Rule = serde_yaml::from_str(&data).unwrap();

	assert_eq!(actual, expected);
}

#[test]
fn serialize_these_requirements() {
	let data = Rule {
		given: Given::TheseRequirements {
			requirements: vec![
				req_ref::Rule {
					requirement: "A Name 1".to_string(),
					optional: false,
				},
				req_ref::Rule {
					requirement: "A Name 2".to_string(),
					optional: true,
				},
			],
			what: GivenCoursesWhatOptions::Courses,
		},
		limit: Some(vec![]),
		filter: Some(filter::Clause::new()),
		action: "count > 2".parse().unwrap(),
	};

	assert_debug_snapshot_matches!("ser_these_reqs", data);
}

#[test]
fn deserialize_these_requirements() {
	let expected = Rule {
		given: Given::TheseRequirements {
			requirements: vec![
				req_ref::Rule {
					requirement: "A Name 1".to_string(),
					optional: false,
				},
				req_ref::Rule {
					requirement: "A Name 2".to_string(),
					optional: true,
				},
			],
			what: GivenCoursesWhatOptions::Courses,
		},
		limit: Some(vec![]),
		filter: Some(filter::Clause::new()),
		action: "count > 2".parse().unwrap(),
	};

	let data = r#"---
given: these-requirements
requirements:
  - requirement: A Name 1
  - {requirement: A Name 2, optional: true}
what: courses
limit: []
where: {}
do: count > 2"#;

	let actual: Rule = serde_yaml::from_str(&data).unwrap();

	assert_eq!(actual, expected);

	let data = r#"---
given: these-requirements
requirements:
  - requirement: A Name 1
  - {requirement: A Name 2, optional: true}
what: courses
limit: []
where: {}
do:
  lhs: Count
  op: GreaterThan
  rhs:
    Integer: 2"#;

	let actual: Rule = serde_yaml::from_str(&data).unwrap();

	assert_eq!(actual, expected);
}

#[test]
fn serialize_areas() {
	let data = Rule {
		given: Given::Areas {
			what: GivenAreasWhatOptions::Areas,
		},
		limit: Some(vec![]),
		filter: Some(filter::Clause::new()),
		action: "count > 2".parse().unwrap(),
	};

	assert_debug_snapshot_matches!("ser_areas", data);
}

#[test]
fn deserialize_areas() {
	let expected = Rule {
		given: Given::Areas {
			what: GivenAreasWhatOptions::Areas,
		},
		limit: Some(vec![]),
		filter: Some(filter::Clause::new()),
		action: "count > 2".parse().unwrap(),
	};

	let data = r#"---
given: areas
what: areas
limit: []
where: {}
do: count > 2"#;

	let actual: Rule = serde_yaml::from_str(&data).unwrap();

	assert_eq!(actual, expected);

	let data = r#"---
given: areas
what: areas
limit: []
where: {}
do:
  lhs: Count
  op: GreaterThan
  rhs:
    Integer: 2"#;

	let actual: Rule = serde_yaml::from_str(&data).unwrap();

	assert_eq!(actual, expected);
}

#[test]
fn serialize_save() {
	let data = Rule {
		given: Given::NamedVariable {
			save: String::from("$my_var"),
			what: GivenCoursesWhatOptions::Courses,
		},
		limit: Some(vec![]),
		filter: Some(filter::Clause::new()),
		action: "count > 2".parse().unwrap(),
	};

	assert_debug_snapshot_matches!("ser_save", data);
}

#[test]
fn deserialize_save() {
	let expected = Rule {
		given: Given::NamedVariable {
			save: String::from("$my_var"),
			what: GivenCoursesWhatOptions::Courses,
		},
		limit: Some(vec![]),
		filter: Some(filter::Clause::new()),
		action: "count > 2".parse().unwrap(),
	};

	let data = r#"---
given: save
save: $my_var
what: courses
limit: []
where: {}
do: count > 2"#;

	let actual: Rule = serde_yaml::from_str(&data).unwrap();

	assert_eq!(actual, expected);

	let data = r#"---
given: save
save: $my_var
what: courses
limit: []
where: {}
do:
  lhs: Count
  op: GreaterThan
  rhs:
    Integer: 2"#;

	let actual: Rule = serde_yaml::from_str(&data).unwrap();

	assert_eq!(actual, expected);
}

#[test]
fn deserialize_save_ba_interim() {
	let data = r#"---
given: save
save: $interim_courses
what: courses
do: count >= 3"#;

	let expected = Rule {
		given: Given::NamedVariable {
			save: String::from("$interim_courses"),
			what: GivenCoursesWhatOptions::Courses,
		},
		limit: None,
		filter: None,
		action: "count >= 3".parse().unwrap(),
	};

	let actual: Rule = serde_yaml::from_str(&data).unwrap();

	assert_eq!(actual, expected);
}

#[test]
fn deserialize_filter_gereqs_single() {
	let data = r#"{where: {gereqs: 'FYW'}, given: courses, what: courses, do: count > 1}"#;

	let expected: filter::Clause = btreemap! {
		"gereqs".into() => value::WrappedValue::eq_string("FYW"),
	};
	let expected = Rule {
		given: Given::AllCourses {
			what: GivenCoursesWhatOptions::Courses,
		},
		limit: None,
		filter: Some(expected),
		action: "count > 1".parse().unwrap(),
	};

	let actual: Rule = serde_yaml::from_str(&data).unwrap();
	assert_eq!(actual, expected);
}

#[test]
fn deserialize_filter_gereqs_or() {
	use value::{TaggedValue, WrappedValue};
	let data = r#"{where: {gereqs: 'MCD | MCG'}, given: courses, what: courses, do: count > 1}"#;

	let expected: filter::Clause = btreemap! {
		"gereqs".into() => WrappedValue::Or(vec![
			TaggedValue::eq_string("MCD"),
			TaggedValue::eq_string("MCG"),
		]),
	};
	let expected = Rule {
		given: Given::AllCourses {
			what: GivenCoursesWhatOptions::Courses,
		},
		limit: None,
		filter: Some(expected),
		action: "count > 1".parse().unwrap(),
	};

	let actual: Rule = serde_yaml::from_str(&data).unwrap();
	assert_eq!(actual, expected);
}

#[test]
fn deserialize_filter_level_gte() {
	use value::{SingleValue::Integer, TaggedValue::GreaterThanEqualTo, WrappedValue::Single};
	let data = r#"{where: {level: '>= 200'}, given: courses, what: courses, do: count > 1}"#;

	let expected: filter::Clause = btreemap! {
		"level".into() => Single(GreaterThanEqualTo(Integer(200)))
	};
	let expected = Rule {
		given: Given::AllCourses {
			what: GivenCoursesWhatOptions::Courses,
		},
		limit: None,
		filter: Some(expected),
		action: "count > 1".parse().unwrap(),
	};

	let actual: Rule = serde_yaml::from_str(&data).unwrap();
	assert_eq!(actual, expected);
}

#[test]
fn deserialize_filter_graded_bool() {
	use value::{SingleValue::Bool, TaggedValue::EqualTo, WrappedValue::Single};
	let data = r#"{where: {graded: 'true'}, given: courses, what: courses, do: count > 1}"#;

	let expected: filter::Clause = btreemap! {
		"graded".into() => Single(EqualTo(Bool(true))),
	};
	let expected = Rule {
		given: Given::AllCourses {
			what: GivenCoursesWhatOptions::Courses,
		},
		limit: None,
		filter: Some(expected),
		action: "count > 1".parse().unwrap(),
	};

	let actual: Rule = serde_yaml::from_str(&data).unwrap();
	assert_eq!(actual, expected);
}

#[test]
fn pretty_print_inline() {
	let input: Rule = serde_yaml::from_str(&"{given: courses, what: courses, do: count >= 1}").unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"take at least one course");
}

#[test]
fn pretty_print_inline_filters() {
	let s = "{given: courses, where: {gereqs: FOL-C}, what: courses, do: count >= 1}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"take at least one course with the “FOL-C” general education attribute");

	let s = "{given: courses, where: {gereqs: SPM}, what: distinct-courses, do: count >= 2}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"take at least two distinct courses with the “SPM” general education attribute");
}

#[test]
fn pretty_print_inline_repeats() {
	let s = "{given: these-courses, repeats: all, courses: [THEAT 233], what: courses, do: count >= 1}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"take THEAT 233 at least one time");

	let s = "{given: these-courses, repeats: all, courses: [THEAT 233], what: courses, do: count >= 4}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"take THEAT 233 at least four times");

	let s = "{given: these-courses, repeats: all, courses: [THEAT 233, THEAT 253], what: courses, do: count >= 4}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"take a combination of THEAT 233 and THEAT 253 at least four times");

	let s = "{given: these-courses, repeats: all, courses: [AMST 205, AMST 206, AMST 207, AMST 208, AMST 209, AMST 210], what: courses, do: count >= 1}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @r###"take at least one of the following courses:

- AMST 205
- AMST 206
- AMST 207
- AMST 208
- AMST 209
- AMST 210"###);

	let s = "{given: these-courses, repeats: all, courses: [THEAT 233], what: credits, do: sum >= 4}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"take THEAT 233 enough times to yield at least four credits");

	let s = "{given: these-courses, repeats: first, courses: [THEAT 233], what: courses, do: count >= 1}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"take THEAT 233");

	let s = "{given: these-courses, repeats: first, courses: [THEAT 233, THEAT 253], what: courses, do: count >= 1}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"take either THEAT 233 or THEAT 253");

	let s = "{given: these-courses, repeats: first, courses: [THEAT 233, THEAT 253], what: courses, do: count >= 2}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"take both THEAT 233 and THEAT 253");
}

#[test]
fn pretty_print_inline_credits() {
	let s = "{given: courses, where: {gereqs: FOL-C}, what: credits, do: sum >= 1}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"have enough courses taken with the “FOL-C” general education attribute to obtain at least one credit");

	let s = "{given: courses, where: {semester: Interim}, what: credits, do: sum >= 3}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"have enough courses taken during Interim semesters to obtain at least three credits");

	let s = "{given: courses, where: {semester: Fall}, what: credits, do: sum >= 10}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"have enough courses taken during Fall semesters to obtain at least ten credits");

	let s = "{given: courses, where: {semester: Fall | Interim}, what: credits, do: sum >= 10}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"have enough courses taken during a Fall or Interim semester to obtain at least ten credits");

	let s = "{given: courses, where: {year: '2012'}, what: credits, do: sum >= 3}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"have enough courses taken during the 2012-13 academic year to obtain at least three credits");

	let s = "{given: courses, where: {institution: St. Olaf College}, what: credits, do: sum >= 17}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"have enough courses taken at St. Olaf College to obtain at least 17 credits");
}

#[test]
fn pretty_print_inline_departments() {
	let s = "{given: courses, where: {gereqs: FOL-C}, what: departments, do: count >= 2}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"have enough courses taken with the “FOL-C” general education attribute to span at least two departments");

	let s = "{given: courses, where: {semester: Interim}, what: departments, do: count >= 1}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"have enough courses taken during Interim semesters to span at least one department");

	let s = "{given: courses, where: { department: '! MATH' }, what: departments, do: count >= 2}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"have enough courses taken outside of the MATH department to span at least two departments");
}

#[test]
fn pretty_print_inline_grades() {
	let s = "{given: courses, where: { gereqs: FOL-C }, what: grades, do: average >= 2.0}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"maintain an average GPA at or above 2.00 from courses taken with the “FOL-C” general education attribute");

	let s = "{given: courses, where: { semester: Interim }, what: grades, do: average >= 3.0}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"maintain an average GPA at or above 3.00 from courses taken during Interim semesters");
}

#[test]
fn pretty_print_inline_terms() {
	let s = "{given: courses, where: { gereqs: FOL-C }, what: terms, do: count >= 2}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"have taken enough courses with the “FOL-C” general education attribute to span at least two terms");

	let s = "{given: courses, where: { semester: Interim }, what: terms, do: count >= 3}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"have taken enough courses during Interim semesters to span at least three terms");
}

#[test]
fn pretty_print_inline_given_requirements_what_courses() {
	let s = "{given: these-requirements, requirements: [{requirement: Core}, {requirement: Modern}], what: credits, do: sum >= 1}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @r###"have the following be true:

1. given the results of the “Core” and “Modern” requirements,
2. there must be at least one credit
"###);

	let s = "{given: these-requirements, requirements: [{requirement: Core}, {requirement: Modern}], where: { gereqs: FOL-C }, what: credits, do: sum >= 1}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @r###"have the following be true:

1. given the results of the “Core” and “Modern” requirements,
2. restricted to only courses taken with the “FOL-C” general education attribute,
3. there must be at least one credit
"###);

	let s = "{given: these-requirements, requirements: [{requirement: Core}, {requirement: Modern}], where: { semester: Interim }, what: credits, do: sum >= 3}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @r###"have the following be true:

1. given the results of the “Core” and “Modern” requirements,
2. restricted to only courses taken during Interim semesters,
3. there must be at least three credits
"###);

	let s = "{given: these-requirements, requirements: [{requirement: Core}, {requirement: Modern}], where: { semester: Fall }, what: credits, do: sum >= 10}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @r###"have the following be true:

1. given the results of the “Core” and “Modern” requirements,
2. restricted to only courses taken during Fall semesters,
3. there must be at least ten credits
"###);

	let s = "{given: these-requirements, requirements: [{requirement: Core}, {requirement: Modern}], where: { year: '2012' }, what: credits, do: sum >= 1}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"have the following be true:

1. given the results of the “Core” and “Modern” requirements,
2. restricted to only courses taken during the 2012-13 academic year,
3. there must be at least one credit
");

	let s = "{given: these-requirements, requirements: [{requirement: Core}, {requirement: Modern}], where: { institution: St. Olaf College }, what: credits, do: sum >= 17}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @r###"have the following be true:

1. given the results of the “Core” and “Modern” requirements,
2. restricted to only courses taken at St. Olaf College,
3. there must be at least 17 credits
"###);
}
