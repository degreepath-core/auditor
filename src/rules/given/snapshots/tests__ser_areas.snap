---
created: "2019-03-09T18:53:39.287040Z"
creator: insta@0.6.3
source: src/rules/given/tests.rs
expression: data
---
Rule(
  given: "areas",
  what: areas({
    "count": Single(GreaterThan(2)),
  }),
  where: Some(AreaClause(
  )),
)
